rfcs/20230606-topk.md

```diff
@@ -0,0 +1,88 @@
+
+## Summary
+
+We propose to add a new op: `stablehlo.topk`, which returns the first `k`
+elements of a tensor and their indices, based on a user-provided comparator. The
+goal of the new op is to bridge a gap between existing and emerging uses of TopK
+in models (especially transformers and LLMs, e.g.: [^1], [^2]), as well as in
+the IR lowering chain.
+
+## Background and Motivation
+
+The TopK operation is ubiquitous in both ML frameworks (e.g.,
+[JAX](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.top_k.html),
+[tensorflow](https://www.tensorflow.org/api_docs/python/tf/math/top_k),
+[PyTorch](https://pytorch.org/docs/stable/generated/torch.topk.html)) and
+compiler IRs ([XLA](https://www.tensorflow.org/xla/operation_semantics#top-k),
+[CHLO](https://github.com/openxla/stablehlo/blob/6f0d25ef2e05f4549c5e4ffc506436a4a13af51f/stablehlo/dialect/ChloOps.td#L811-L842),
+[MHLO](https://github.com/openxla/xla/commit/11a4424760f8814faa6da29a0e6dd9d5c168d22a#diff-9573e4e3a2bf686edee3437726f9dbccaf1b296e634d44f4df504d94bf2576d1),
+[LinalgExt](https://github.com/openxla/iree/blob/b5e45cd08eecbd97b0568f9c5fdac28b0f5e8f11/llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td#L436-L480)),
+which makes its absence from StableHLO a point of friction. For example, in IREE
+we have to pattern match stablehlo `iota` + `sort` + `slice` and raise it back
+to `chlo.top_k`, so that we can efficiently lower it to `linalg_ext.topk`. See
+the IREE issue for more details: <https://github.com/openxla/iree/issues/13905>.
+
+The definitions of TopK in frontend frameworks and IRs vary in two places:
+
+* Whether `k` is static or dynamic.
+* Whether the results are sorted, unordered, or whether ordering is decided based
+  on a function argument.
+
+Note that when the input TopK gets lowered to `sort` followed by a dynamic slice,
+we generally lose the information about both. If `k` is a runtime value, we no
+longer know whether we can expect it to be much smaller than the number of
+elements, which makes it difficult to tell whether to rewrite sort as TopK or
+whether we truly have to sort all elements. Second, we may be introducing
+unnecessary ordering when the input semantics do not require for the results to
+be sorted.
+
+Note that we want to avoid degrading to a general `sort`, as this can lead to
+significant bottlenecks in real-world models, e.g.:
+<https://github.com/openxla/iree/issues/13865>.
+
+This loss of high level semantic information and high friction along the
+lowering chain makes us argue for a dedicated TopK op in StableHLO.
+
+## Implementation
+
+We propose to add a `stablehlo.topk` op with dynamic argument `k` and an
+attribute `order` (either `SORTED` or `UNORDERED`). This version of TopK should
+be a faithful representation of TopK in input frameworks that does not introduce
+needless sorting.
+
+Sample IR:
+
+```mlir
+%values, %indices = stablehlo.topk %values, %k, UNORDERED {
```

Comment:
small nit: Why an attribute versus a boolean arg?

Reply:
In all of the uses of topk in frameworks I saw, the ordering was decided statically for each callsite. If dynamism is required, it should be simple to emit a select with two variants.

Reply:
I added this to 'Additional Considerations'